triSYCL with Eigen compatibility

a-doumoulakis opened the pull request:
When trying to make the Eigen tests work with triSYCL I encountered several problems, in this pull request are the changes required to (mostly) fix those problems. I also updated the Eigen build files to be able to use triSYCL instead of ComputeCPP as the SYCL implementation. The [branch](https://bitbucket.org/a-doumoulakis/opencl "Bitbucket") containing the updated build files should be merged soon with the Eigen master.

--------------

## Notes and summary of changes

### Buffer
- Added the ability to give a custom allocator to a sycl buffer
> The get_destructor_future method of private::buffer had to be made public because it is not accessible from buffer::waiter if it was instantiated with a custom allocaltor. 

### Task
- Added a lock to protect the write of ```execution_ended``` to address deadlocks caused by a race condition between the modification of ```execution_ended```, the call to ```notify_all``` and ```ready.wait``` in task.hpp. This was raised in #37. 

### Handler
- Fixed a template for ```parallel_for``` that did not have a default kernel name

### Parallelism
- Added OpenMP parallelism with the parallel_for function that takes an ```nd_range``` as an argument. The current implementation creates one thread per work-item. This is needed for the barriers used in Eigen that need to synchronize all the thread in a work-group.
- Modified ```parallel_for_workitem```, this method is used to parallelize a single work-group, this will have to be better implemented at some point. 

### Misc
- Created *include/SYCL/sycl.hpp* to enable ```#include<SYCL/sycl>``` used in Eigen tests.
- Added wait to ```wait_and_throw```
- Typos

keryell commented on include/CL/sycl/buffer/detail/buffer.hpp:
```diff
@@ -349,29 +349,12 @@ class buffer : public detail::buffer_base,
 
       \return A \c future in the \c optional if there is something to
       wait for, otherwise an empty \c optional
+      \todo Make the function private again
```
Instead of moving the code around it is simpler to understand history by just adding a
```
public:
```
at the right place.

keryell commented on include/CL/sycl/device.hpp:
```diff
@@ -248,13 +248,13 @@ class device
 template <>
 inline auto device::get_info<info::device::max_compute_units>() const {
-  return size_t { 56 };
+  return size_t { 8 };
```
And now with these 2 previous changes this unit test is failing:
```
/home/keryell/Xilinx/Projects/OpenCL/SYCL/triSYCL/branch/tmp/tests/device/get_info.cpp(14): test d.get_info<info::device::max_work_group_size>() == 63 failed in function: 'int test_main(int, char**)'
``` 
and so on.
Please update the test too.

keryell commented on include/CL/sycl/parallelism/detail/parallelism.hpp:
```diff
@@ -294,44 +313,30 @@ void parallel_for_workitem(const group<Dimensions> &g,
+    nd_item<Dimensions> index { g.get_nd_range() };
+    id<Dimensions> local; // to initialize correctly
+#pragma omp for nowait
+    for (int th_id = 0; th_id < tot; ++th_id) {
```
I changed this recently to
```
for (std::size_t th_id = 0; th_id != tot; ++th_id) 
```
to remove a compilation warning.
So I guess you could keep my loop definition instead.

keryell commented on include/CL/sycl/parallelism/detail/parallelism.hpp:
```diff
@@ -294,44 +313,30 @@ void parallel_for_workitem(const group<Dimensions> &g,
+      } else if (Dimensions == 3) {
+        local[0] = th_id / (l_r.get(1)*l_r.get(2));
+        local[1] = (th_id / l_r.get(2)) % l_r.get(1);
+        local[2] = th_id % l_r.get(2);
```
To be simplified later in another PR according to https://github.com/triSYCL/triSYCL/issues/43 some day...

keryell commented on include/SYCL/sycl.hpp:
```diff
@@ -0,0 +1 @@
+#include "../CL/sycl.hpp"
```
Do not add this.

It is a ComputeCpp bug which has been fixed long time ago.
Make a PR against Eigen & TensorFlow instead. :-)

a-doumoulakis commented:
I changed the *get_info* test to check for the new returned sizes and all test succeed. 

I also modified the type of *th_id* in *parallelism.hpp* to suppress the warning and also did the same thing for the Jacobi tests while I was at it. I had to keep the condition in the loop as ```th_id < tot``` because OpenMP gives the error *Invalid control predicate* with ```th_id != tot ```

Other than that I removed SYCL/sycl.hpp and changed the included header in Eigen instead.

keryell commented:
Thank you for your contribution!
